clientapi/routing/room_tagging.go

```diff
@@ -0,0 +1,265 @@
+	tagContent := newTag()
+	err = json.Unmarshal(dataByte, &tagData)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	err = json.Unmarshal(tagData[0].Content, &tagContent)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// send data to syncapi
+	if err := syncProducer.SendData(userID, roomID, "m.tag"); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: tagContent,
+	}
+}
+
+// PutTag implements PUT /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+// Put functionality works by getting existing data from the DB (if any), adding
+// the tag onto the "map" and saving the new "map" onto the DB
+func PutTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	device *authtypes.Device,
+	userID string,
+	roomID string,
+	tag string,
+	syncProducer *producers.SyncAPIProducer,
+) util.JSONResponse {
+
+	if device.UserID != userID {
+		return util.JSONResponse{
+			Code: http.StatusForbidden,
+			JSON: jsonerror.Forbidden("Cannot modify another user's tags"),
+		}
+	}
+
+	localpart, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	var properties gomatrix.TagProperties
+
+	if reqErr := httputil.UnmarshalJSONRequest(req, &properties); reqErr != nil {
+		return *reqErr
+	}
+
+	tagContent := newTag()
+	var dataByte []byte
+	if len(data) > 0 {
+		dataByte, err = json.Marshal(data)
+		if err != nil {
+			return httputil.LogThenError(req, err)
+		}
+		if err = json.Unmarshal(dataByte, &tagContent); err != nil {
+			return httputil.LogThenError(req, err)
+		}
+	}
+	tagContent.Tags[tag] = properties
+	err = saveTagData(req, localpart, roomID, accountDB, tagContent)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// send data to syncapi
+	if err := syncProducer.SendData(userID, roomID, "m.tag"); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: struct{}{},
+	}
+}
+
+// DeleteTag implements DELETE /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+// Delete functionality works by obtaining the saved Tags, removing the intended tag from
+// the "map" and then saving the new "map" in the DB
+func DeleteTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	device *authtypes.Device,
+	userID string,
+	roomID string,
+	tag string,
+	syncProducer *producers.SyncAPIProducer,
+) util.JSONResponse {
+
+	if device.UserID != userID {
+		return util.JSONResponse{
+			Code: http.StatusForbidden,
+			JSON: jsonerror.Forbidden("Cannot delete another user's tags"),
+		}
+	}
+
+	localpart, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// If there are no tags in the database, exit.
+	if len(data) == 0 {
+		//Synapse returns a 200 OK response on finding no Tags, same policy is followed here.
```

Comment:
Really we do this because [the spec doesn't specify](https://matrix.org/docs/spec/client_server/unstable#delete-matrix-client-r0-user-userid-rooms-roomid-tags-tag) any return code other than 200, rather than following what Synapse does.

Reply:
Ah alright. I remembered something related to synapse in the last discussion.